src/cargo/util/graph.rs

```diff
@@ -154,17 +324,251 @@ impl<N: fmt::Display + Eq + Ord + Clone, E: Clone> fmt::Debug for Graph<N, E> {
     }
 }
 
-impl<N: Eq + Ord + Clone, E: Eq + Clone> PartialEq for Graph<N, E> {
+impl<N: Eq + Hash + Clone, E: Eq + Clone + PartialEq> PartialEq for Graph<N, E> {
     fn eq(&self, other: &Graph<N, E>) -> bool {
         self.nodes.eq(&other.nodes)
     }
 }
-impl<N: Eq + Ord + Clone, E: Eq + Clone> Eq for Graph<N, E> {}
+impl<N: Eq + Hash + Clone, E: Eq + Clone> Eq for Graph<N, E> {}
 
-impl<N: Eq + Ord + Clone, E: Clone> Clone for Graph<N, E> {
-    fn clone(&self) -> Graph<N, E> {
-        Graph {
-            nodes: self.nodes.clone(),
+#[derive(Clone, Debug)]
+pub struct Edges<'a, E: Clone> {
+    graph: &'a [EdgeLink<E>],
+    index: Option<EdgeIndex>,
+}
+
+impl<'a, E: Clone> Edges<'a, E> {
+    pub fn is_empty(&self) -> bool {
+        self.index
+            .and_then(|idx| self.graph.get(idx))
+            .and_then(|l| l.value.as_ref())
+            .is_none()
+    }
+}
+
+impl<'a, E: Clone> Iterator for Edges<'a, E> {
+    type Item = &'a E;
+
+    fn next(&mut self) -> Option<&'a E> {
+        while let Some(edge_link) = self.index.and_then(|old_index| self.graph.get(old_index)) {
```

Comment:
Is the `get` here required to be a fallible lookup? I'd imagine that if we're iterating all internal edge pointers should already be valid